File: QuSoC/QuSoC/apps/Arrays/firmware/Arrays.cpp
#include "Arrays.h"
#include <stdint.h>
#include <stdbool.h>
#include "soc.h"
namespace Arrays
{
	int8_t Firmware::S8Buff[16] = {0};
	int16_t Firmware::S16Buff[16] = {0};
	uint32_t Firmware::U32Buff[16] = {0};
	void Firmware::Fill(uint32_t size)
	{
		uint32_t offset = (3 + ((size >> 1)));
		for(uint32_t i = 0; (i < size); (i++))
		{
			uint32_t value = (i - offset);
			S8Buff[i] = (int8_t)value;
			S16Buff[i] = (int16_t)value;
			U32Buff[i] = value;
		}
	}
	void Firmware::Sum(uint32_t size)
	{
		uint32_t result = 42;
		for(uint32_t i = 0; (i < size); (i++))
		{
			result += (uint32_t)S8Buff[i];
			result += (uint32_t)S16Buff[i];
			result += U32Buff[i];
		}
		Arrays_SOC_Counter = result;
	}
	void Firmware::EntryPoint()
	{
		uint32_t size = 6;
		Fill(size);
		Sum(size);
	}
}

File: QuSoC/QuSoC/apps/Arrays/source/Arrays.cs
namespace Arrays
{
    public class SOC
    {
        public static SOC Instance { get; set; } = new SOC();

        public virtual uint Counter { get; set; }
    }

    public static class Firmware
    {
        static sbyte[] S8Buff = new sbyte[16];
        static short[] S16Buff = new short[16];
        static uint[] U32Buff = new uint[16];

        static void Fill(uint size)
        {
            uint offset = 3 + (size >> 1);
            for (uint i = 0; i < size; i++)
            {
                var value = i - offset;
                S8Buff[i] = (sbyte)value;
                S16Buff[i] = (short)value;
                U32Buff[i] = value;
            }
        }

        static void Sum(uint size)
        {
            uint result = 42;
            for (uint i = 0; i < size; i++)
            {
                result += (uint)S8Buff[i];
                result += (uint)S16Buff[i];
                result += U32Buff[i];
            }

            SOC.Instance.Counter = result;
        }

        public static void EntryPoint()
        {
            uint size = 6;
            Fill(size);
            Sum(size);
        }
    }
}
